### ISSUE TITLE
pmem_allocator doesn't work with pmem_is_pmem

### ISSUE THREAD

**Issue by meng-sun** [memkind/memkind#342]
Hi, I've been trying to use the pmem_allocator API with STL vector. When I tried running timers on it the runtimes were very similar between pmem_allocator and the default STL vector, so I tried to test if it was allocating pmem properly with pmem_is_pmem. I get 0 from this code snippet:

```
#include <libpmem.h>
#include <cstdlib>
#include <pmem_allocator.h>
#include <vector>
#include <cassert>

int main() {
  libmemkind::pmem::allocator<int> alc{ "/path/to/dir/", MEMKIND_PMEM_MIN_SIZE };
  std::vector<int, libmemkind::pmem::allocator<int>> v{ alc };
  for (int i = 0; i < 20; ++i) {
      v.push_back(0xDEAD + i);
      assert(v.back() == 0xDEAD + i);
  }
  std::cout << "Is it pmem " << pmem_is_pmem(&v.front(), 4) << std::endl;
  return 0;
}
```
Is pmem_allocator not compatible with pmem_is_pmem? I verified that pmem_is_pmem is working for mmap'd pmem using libpmem. If so, is there any pmem_allocator-compatible check? Tagging @jmftrindade.

Thanks.

**Comment by michalbiesek** [comment]
Hi,
Thanks for your message.
Unfortunately, **libpmem** is not compatible with **libmemkind**.
Libmemkind with memkind.h interface provides function **memkind_detect_kind** which could be used to verify if allocation belongs to a specific kind, therefore it could be used to check if the allocated memory belongs to pmem (see https://github.com/memkind/memkind/blob/master/examples/pmem_detect_kind.c for details)
Please, be aware there is one limitation with this function - it works correctly only with allocation which was done by libmemkind API - in other cases results with Undefined Behaviour.

**Comment by meng-sun** [comment]
Sorry, I have some confusion regarding the link. Memkind_detect_kind requires a pointer to memory and then a flag for the pmem type. I found a AllocatorTypes::MEMKIND_PMEM but it's included in the tests config file. Is there a libmemkind::kinds value for PMEM?

I'm trying to test whether or not my vector is probably allocated on NVRAM in app-direct mode.

**Comment by meng-sun** [comment]
Nevermind, I'm good, closing the issue. Thanks.

**Comment by meng-sun** [comment]
Hi, an update. I'm still trying to find a guaranteed method for verifying that the allocated values are on pmem. I have run a variant of the previous code but with ~600K ints through strace and I see this

```
openat(AT_FDCWD, "/mnt/pmem/memkind.QAp2Ey", O_RDWR|O_CREAT|O_EXCL, 0600) = 3
unlink("/mnt/pmem/memkind.QAp2Ey")    = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
futex(0x7f17c906b7d4, FUTEX_WAKE_PRIVATE, 2147483647) = 0
mmap(NULL, 6291456, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_NORESERVE, -1, 0) = 0x7f17c5e00000
```

and then a bunch of anonymous maps of 2 or 8MB, then

```
futex(0x7f17c6c1d24c, FUTEX_WAKE_PRIVATE, 2147483647) = 0
fallocate(3, 0, 0, 29360128)            = 0
mmap(NULL, 29360128, PROT_READ|PROT_WRITE, MAP_SHARED, 3, 0) = 0x7f17a3800000
mmap(NULL, 10485760, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_NORESERVE, -1, 0) = 0x7f17a2e00000
```

So it seems like it is grabbing pmem, so then I tried to make ints using the allocator and view them later (to ensure that the objects are actually getting written to pmem). I removed the unlink in the memkind_tmpfile struct, and when I viewed the file I have a bunch of null bytes. This might be because it doesn't flush to pmem, but I'm not clear on how the allocator works.

What is the correct way of doing this verification? Thanks a lot.

**Comment by michalbiesek** [comment]
Hi,
Thanks for your message.
I am not sure if I understand you correctly.
**pmemallocator** to which you refer is basically a wrapper for other functions from libmemkind  which allows:
- creating pmem _kind_ object (by calling **memkind_create_pmem** )
- allocate/deallocate memory using pmem _kind_ by (void ***memkind_malloc**(memkind_t _kind_, size_t size / void **memkind_free**(memkind_t _kind_, void *ptr);)

One of advantage of **pmemallocator** is that you don't have to remember and pass _kind_ structure because the _kind_ structure is kept in **pmemallocator** internal data.
Allocation using **pmemallocator** is based on file-backed on top of a temporary file in the given directory. The directory which should point mount path with DAX filesystem.
Libmemkind doesn't verify if path point to DAX filesystem, for details, see discussion in #107

Additionally, memkind is built on top of jemalloc - therefore jemalloc itself is responsible for heap management. In other words, even for correct allocation to PMEM it uses DRAM to keep some metadata.

**Comment by jmftrindade** [comment]
Hey @michalbiesek , thanks for the prompt response.  Let me see if I can help clarify the miscommunication here :)

@meng-sun has been trying to use memkind for dynamic allocation onto NVRAM with code that is similar to the following:

```cpp
  libmemkind::pmem::allocator<int> alc{ "/path/to/dir/", MEMKIND_PMEM_MIN_SIZE };
  std::vector<int, libmemkind::pmem::allocator<int>> v{ alc };
```

However, @meng-sun is finding that the latency of reads for this example above is essentially the same as that of **not using** memkind's `pmem::allocator` at all. That is, accessing elements of a DRAM allocated `std::vector<int>` is no faster than accessing elements of a NVRAM allocated `std::vector<int, libmemkind::pmem::allocator<int>>`, which should not be the case.

So the most pressing questions @meng-sun has are:

1. Is @meng-sun code snippet using `pmem::allocator` the right way to use memkind to allocate dynamic memory of a `std::vector` onto NVRAM?
2. What is the correct way to programmatically double check that dynamic memory allocated with `pmem::allocator` has indeed been allocated onto NVRAM hardware that we have available on our server?

Thanks!
Joana

**Comment by michalbiesek** [comment]
Hi Joana,
Thanks for chiming in. Let me answer your questions:

_Is @meng-sun code snippet using pmem::allocator the right way to use memkind to allocate dynamic memory of a std::vector onto NVRAM?_
Yes, the usage of pmem::allocator from your example looks fine

_What is the correct way to programmatically double-check that dynamic memory allocated with pmem::allocator has indeed been allocated onto NVRAM hardware that we have available on our server?_
If you want to ensure that you have correctly configured NVRAM hardware - you could use libpmemobj to verify that.

E.g. you could verfiy the path by using something like:

`pmem_map_file(file_path_to_verify, size, PMEM_FILE_CREATE, 0666, &mapped_len, &is_pmemp);`

and check is_pmemp variable, according to libpmemobj doc "If is_pmemp is not NULL, a flag indicating whether the mapped file is actual pmem"
Be aware that in memkind you pass the path to a directory, wherein example above the path point to a directory+file.

Just to ensure you configured NVRAM hardware like from this [instruction](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/storage_administration_guide/configuring-persistent-memory-for-file-system-direct-access-dax)?

Regarding your latency measure could you share more details about your methodology? Was it multithread? What was the size of the objects?

To verify that you allocate in NVRAM - you could use `df -h` command to check how much memory you use in the mounted point of NVRAM hardware during your test.

**Comment by meng-sun** [comment]
Hi Michal, the is_pmemp doesn't return NULL for the mounted path. Also, the amount of memory used by the program (measured through df -h) matches the expected amount (amount allocated by the allocater.)

Regarding the methodology, we compare both a multithreaded and singlethreaded version of a program that reads objects, does a logic check on each object, and returns a vector of references to objects that passed the logic check. In one version, all data (1M objects of 48 bytes each) is stored on NVRAM, and in the other, it's all stored on DRAM. No optimizations and with cache cleared at the start. We are seeing about a 5-7% difference between NVRAM and DRAM (NVRAM being slower) and a total runtime of ~25s for the single threaded and ~7s for the multithreaded (4 threads) version.

I also ran this single threaded test:
```
int main() {
  timer d_timer;
  timer n_timer;

  size_t max_size = 1024 * 100000;
  libmemkind::pmem::allocator<int> alc{ "/path/to/pmem", 5*max_size};
  std::vector<int, libmemkind::pmem::allocator<int>> v{ alc };
  std::vector<int> v_dram;
  v.resize(max_size);
  v_dram.resize(max_size);

  for (int i=0; i<max_size; i+=1) {
    d_timer.start();
    v_dram[i] = rand();
    d_timer.stop();
    n_timer.start();
    v[i] = rand();
    n_timer.stop();
  }
```
But here the total dram is significantly slower (~0.2s) than the total nvram access tune?

Also additional info: 
System -- ext4 with DAX enabled on mount
Namespace properties -- 
  {
    "dev":"namespace0.0",
    "mode":"fsdax",
    "map":"dev",
    "size":1059076440064,
    "uuid":<some uuid>,
    "sector_size":512,
    "align":2097152,
    "blockdev":"pmem0"
  }

Thanks.

**Comment by kilobyte** [comment]
As for detecting pmem: the best way is to try mmap() with flags MAP_SHARED_VALIDATE|MAP_SYNC — which will succeed only if DAX is supported and enabled for that particular directory.

Other ways to do so are hacks:
* writing something to the last page of a `mmap`ped allocation, `truncate()`ing the underlying file so that last page is only partially backed, truncating it back to full size then seeing if the contents have been preserved.  To avoid an externally-visible information leak, the operating system has to write actual zeroes to the persistent memory — a step that's not needed for a regular disk (or tmpfs).  Thus, persistent memory will return zeroes in that part.
* accessing a page, calling `mincore()`, `madvise(MADV_DONTNEED)`, `mincore()` — anonymous memory will fail the `mincore` call, disk/tmpfs backed memory won't discard the memory immediately, while persistent memory will unfault it instantly.  Alas, this method is racey as a pathologically loaded system might require the madvised page before the second mincore call, but on the other hand, no outside knowledge about the memory you query is needed: no need for a file descriptor, path, etc.

**Comment by kilobyte** [comment]
As for the benchmark:
* You are trying to measure a time that's orders of magnitude shorter than your test harness (two syscalls or RDTSCs!).
* Also, both turning the timers on/off, and calculating rand(), do write to memory, which interferes with the test — the processor can parallelize writes to different memory channels, thus writing A to fast memory "then" B to slightly slower memory can be done in parallel while writing A then B to the same memory chip has to be done one after another.
* But you don't actually write to memory at all — all those stores land in processor's cache, where the processor can then discharge them on its leisure, possibly hours later.  To benchmark the memory rather than processor, you'd need to optimize the loop at least enough to overwhelm the memory bandwidth.
* Unless you drain the stores in some way, the test vector would need to be much larger than processor's cache.  Modern processors have on the order of 64MB of L3, which, while won't fit 390MB you write, holds a non-negligible percentage of the test.

I'd recommend pulling the timer outside the loop:
```
d_timer.start();
for (...)
    v_dram...;
d_timer.stop();
n_timer.start();
for (...)
    v...;
n_timer.stop();
```

No hardware memory does compression (there are some disks that do, also zswap), thus you could get rid of `rand()` and write just anything.  '*' (= 42) is a good 8-bit value, so is 32-bit 0x776f656d. ☺

**Comment by jmftrindade** [comment]
I agree with @kilobyte 's comments above, thanks for chiming in.

In the end I was able to solve @meng-sun's issue by looking at example code included in the library.  The [`STL_MAP_INT_STRING_TEST`](https://github.com/memkind/memkind/blob/f732a39de85648a73a67d25872dc785a3ac1431b/examples/pmem_cpp_allocator.cpp#L117) unit test was particularly useful.
